### DISCUSSION TITLE
Is it possible to include SQLite.Interop.dll for linux-x64 and osx-x64 into linq2db.SQLite nuget package?

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4560] [General]
Hello.
Stub.System.Data.SQLite.Core.NetStandard nuget package contains SQLite.Interop.dll native library for 4 platforms:
![image](https://github.com/linq2db/linq2db/assets/157357768/6274e8ef-b164-45ea-a3bc-7a637801c913)
While linq2db.SQLite nuget package contains that library only for 2 platforms (I suppose that linux-x64 and osx-x64 are missing):
![image](https://github.com/linq2db/linq2db/assets/157357768/f8c6ce99-ab51-4859-903b-d6e4a6454f3b)
I include SQLite.Interop.dll as content in my **.NET Standard 2.0** class library manually [using](https://stackoverflow.com/questions/32639630/sqlite-interop-dll-files-does-not-copy-to-project-output-path-when-required-by-r/60122304#60122304) `GeneratePathProperty="true"`. I have to do that because I use class library from **.NET Framework 4.7.2** console app - runtimes folder is not copied to output on build (and `connection.LoadExtension("SQLite.Interop.dll", "sqlite3_fts5_init"`); fails with `System.BadImageFormatException`):
![image](https://github.com/linq2db/linq2db/assets/157357768/7c93dbdc-5db7-4336-a8b7-c7236e560839)
I can do that only if I additionally reference `Stub.System.Data.SQLite.Core.NetStandard` nuget but I'd like to use just linq2db.SQLite.
Thank you.

**Comment by MaceWindu** [comment]
Those are used for T4 templates only. Any reason why not use linq2db.cli for scaffold? It was introduced as cross-platform replacement of T4 templates.

I doubt adding non-windows runtimes to T4 will work (with Reader?) as it will use .net framework assembly and it probably don't even know about those runtimes

**Reply by alexey-leonovich** [reply]
I include SQLite.Interop.dll in my project not for database model scaffolding purposes. I need that library to be present in output directory to be able to use FTS5 extesion via:
```
var connection = new SQLiteConnection(connectionString);
connection.Open();
connection.EnableExtensions(true);
connection.LoadExtension("SQLite.Interop.dll", "sqlite3_fts5_init");
return connection;
```

**Reply by MaceWindu** [reply]
`linq2db.<db_name>` nugets mainly required for T4 scaffold functionality. If you don't use it we recommend to reference linq2db and required provider(s) nugets.

Also `tools` folder contains binaries required for T4 and not expected to be used by user directly. `linq2db.Sqlite` already reference `Stub.System.Data.SQLite.Core.NetStandard` as transient dependency https://www.nuget.org/packages/System.Data.SQLite.Core#dependencies-body-tab and I would expect runtimes folder copied to output on build.

**Reply by MaceWindu** [reply]
Looks like I know what the issue here. Let me check how to do it properly

**Reply by MaceWindu** [reply]
Ok, I don't know how to easily load extension from `Sqlite.interop` for now, but I don't think you need it for FTS5. It should be enabled by default.

**Reply by MaceWindu** [reply]
Ok, correct answer is: FTS5 is not enabled, but I had no issue loading it using `cn.LoadExtension("SQLite.Interop.dll", "sqlite3_fts5_init");`

Are you sure you have issues with this code or you just noticed that tools missing runtime? because tools folder is not used by code in any way.

**Reply by MaceWindu** [reply]
Only issue you could have is with ARM platforms as system.data.sqlite currently doesn't provide ARM builds

**Reply by alexey-leonovich** [reply]
> If you don't use it we recommend to reference linq2db and required provider(s) nugets.

Didn't know that, thank you. According this recommendation my project will have `linq2db` and `Stub.System.Data.SQLite.Core.NetStandard` nugets referenced.
> Also `tools` folder contains binaries required for T4 and not expected to be used by user directly. 

I got it. So I will need to use libraries from `Stub.System.Data.SQLite.Core.NetStandard` nuget package cache folder.
> `linq2db.Sqlite` already reference `Stub.System.Data.SQLite.Core.NetStandard` as transient dependency and I would expect runtimes folder copied to output on build.

Runtime folder is copied, but according to System.Data.SQLite [wiki](https://system.data.sqlite.org/index.html/doc/trunk/www/downloads.wiki) SQLite.Interop.dll should be placed into \x86 and \x64 subfolders:
![image](https://github.com/linq2db/linq2db/assets/157357768/81d5b8fa-4ba3-4abe-92a9-029cf08a51f1)

**UPDATE**
> Are you sure you have issues with this code or you just noticed that tools missing runtime?

Surprisingly looks like it's working with current runtime folder structure (no need to have \x86 and \x64). Sorry for misleading.

**Reply by MaceWindu** [reply]
That page describes how it work in old days of .net framework. Now usually build/publish generates deps json file which reference dependencies from nuget cache

**Reply by alexey-leonovich** [reply]
> how it work in old days of .net framework

Now I remember that my problem only appeared on .NET Framework 4.7.2. If I use my **.NET Standard 2.0** class library from a **.NET 8** console app - it's working. If I use it from a **.NET Framework 4.7.2** console app - runtimes folder **is not** copied to output on build (and `connection.LoadExtension("SQLite.Interop.dll", "sqlite3_fts5_init");` fails with BadImageFormatException):
![image](https://github.com/linq2db/linq2db/assets/157357768/7c93dbdc-5db7-4336-a8b7-c7236e560839)
So I needed to add SQLite.Interop.dll as content manually. Are you suppose it to work out of the box in that scenario somehow?

**Reply by MaceWindu** [reply]
I don't observe this issue with net472/net48 target. Do you use sdk-style project (`<Project Sdk="Microsoft.NET.Sdk">`) or legacy one?

**Reply by MaceWindu** [reply]
Legacy project (`<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">`) also works for me

**Reply by alexey-leonovich** [reply]
I use SDK-style project. I've created a [sample repo](https://github.com/alexey-leonovich/LinqToDBTest). Just comlipe and run **TestAppNFW** project. P. S. I'm on Win 10 22H2 x64 Pro.

**Reply by MaceWindu** [reply]
Replace 
```xml
<PackageReference Include="Stub.System.Data.SQLite.Core.NetStandard" Version="1.0.118" />
```
with
```xml
<PackageReference Include="System.Data.SQLite.Core" Version="1.0.118" />
```

**Reply by alexey-leonovich** [reply]
Thank you. It's working now.
